### DISCUSSION TITLE
Renovate not creating PRs to update npm security vulnerabilities by Dependabot

### DISCUSSION THREAD

**Discussion by pbas-bs** [renovatebot/renovate#28451] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I am trying to have Renovate open a PR when Dependabot alerts a security vulnerability.
I created a minimal example here: https://github.com/pbas-bs/renovate-npm-test

I have a security vulnerability reported:
<img width="1053" alt="image" src="https://github.com/renovatebot/renovate/assets/145019302/71c72e14-dd15-4843-8afa-6aa25bde1c4d">

My renovate configuration enables security vulnerabilities and transitive remediation:
```json5
{
  $schema: "https://docs.renovatebot.com/renovate-schema.json",
  extends: [
    "config:base",
    "schedule:weekly",
    ":autodetectPinVersions",
    ":enableVulnerabilityAlerts",
    ":labels(dependencies)",
    ":preserveSemverRanges",
    ":prImmediately",
    ":rebaseStalePrs",
    ":semanticCommits",
    ":semanticPrefixChore",
    ":separateMajorReleases",
    ":timezone(Europe/Rome)",
  ],
  packageRules: [
    {
      groupName: "node minor + patch dependencies",
      matchManagers: ["npm"],
      matchUpdateTypes: ["minor", "patch"],
    },
  ],
  draftPR: true,
  reviewersFromCodeOwners: false,
  reviewersSampleSize: 1,
  stabilityDays: 2,
  vulnerabilityAlerts: {
    enabled: true,
    draftPR: false,
    reviewersFromCodeOwners: true,
  },
  prConcurrentLimit: 3,
  prHourlyLimit: 2,
  enabledManagers: ["npm"],
  transitiveRemediation: true,
}
```

From the logs it appears as Renovate finds the two allerts.
Full logs: https://gist.github.com/pbas-bs/3e49d8f885d19a04ac9a930425788724

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: alert package rules
{
  "alertPackageRules": [
    {
      "matchDatasources": [
        "npm"
      ],
      "matchPackageNames": [
        "nth-check"
      ],
      "matchCurrentVersion": "= 1.0.2",
      "matchFileNames": [
        "package-lock.json"
      ]
    },
    {
      "matchDatasources": [
        "npm"
      ],
      "matchPackageNames": [
        "postcss"
      ],
      "matchCurrentVersion": "= 7.0.39",
      "matchFileNames": [
        "package-lock.json"
      ]
    }
  ]
}
...
DEBUG: 2 flattened updates found: nth-check, postcss
DEBUG: Returning 2 branch(es)
DEBUG: config.repoIsOnboarded=true

```

</details>

**Comment by viceice** [comment, marked as answer]
check `packageFiles with updates` message. renovate can't update transitive npm dependencies

**Reply by pbas-bs** [reply]
Hi! Thanks for the clarification. 

I was confused by the [documentation](https://docs.renovatebot.com/configuration-options/#transitiveremediation):
```
Applicable only for GitHub platform (with vulnerability alerts enabled) and npm manager.
When the lockfileVersion is higher than 1 in package-lock.json, remediations are only possible when changes are made to package.json.
```
To confirm, does this mean that:
1.  `Renovate can't update transitive npm dependencies` but can for other managers (e.g. python)?
2. Or that Renovate can't update transitive dependencies in general?

Thanks for your time!

**Reply by rarkins** [reply]
2

**Reply by pbas-bs** [reply]
All right, thanks!

**Reply by SchroederSteffen** [reply]
I came across this limitation this week and wanted to summarize it for our internal docs.
Could you please let me know if I got it right?

---

According to [this response](https://github.com/renovatebot/renovate/discussions/28451#discussioncomment-9153000), Renovate doesn't support transitive dependency updates in general.

This sounds odd, because there is even the setting [transitiveRemediation](https://docs.renovatebot.com/configuration-options/#transitiveremediation) that seems to enable exactly such updates.
However, the above documentation mentions the following:
> remediations are only possible when changes are made to package.json.

This means that a direct dependency needs to be updated to update transient dependencies.
And updating a direct dependency only helps if that direct dependency expects a higher transient dependency version which lets NPM do the update of the transient dependency.

Summary: Vulnerability alerts for transient dependencies can only be fixed by Renovate if your direct dependency increases its required version to a fixed version of the transient dependency.

**Reply by HonkingGoose** [reply]
> I came across this limitation this week and wanted to summarize it for our internal docs.
Could you please let me know if I got it right?

I'll let a maintainer respond to your summary.

We should have the correct information in the official docs. I don't want users to hack around missing/wrong info in our docs, by putting it in their internal docs. Those internal docs will get outdated, and it's more work for the user to maintain than for us. :smile: 

Once the correct info is known, that should go in the docs. I'm happy to review a docs PR for this. :wink:

**Reply by rarkins** [reply]
The simplest way to put it for now: vulnerability remediations are possible only when they affect a direct dependency. It's possible to get lockfile-only remediations when the affected dependency has a range in package.json

**Reply by SchroederSteffen** [reply]
Thanks for your responses. I'd also prefer it if the official docs would explain it properly instead of having to write own internal docs.

> It's possible to get lockfile-only remediations when the affected dependency has a range in package.json

Sorry but I don't quite understand that. Do you mean that the affected/direct dependency must not be pinned or that it must use a range for the vulnerable transient dependency?
The first contradicts with Renovate's own best practices(?) and the second doesn't seem to work in our case.

**Reply by rarkins** [reply]
The fact I wrote does not contradict our recommendations. Our recommendation is that you pin dependencies unless you're writing a library. The fact I wrote is that if you have unpinned dependencies then you may get a lockfile-only CVE remediation. Example:

- Package file contains `^1.0.0`
- Lock file contains `1.0.1`
- CVE is fixed in `1.0.2`

In this case you would expect to receive a PR updating the lockfile from 1.0.1 to 1.0.2.

**Reply by SchroederSteffen** [reply]
Thanks for the clarification. Got it.
Then, I would suggest to rewrite it as follows:
> It's possible to get lockfile-only remediations when the affected direct dependency is not pinned in your package.json

Right, we also don't pin our libraries' non-dev dependencies. All others are pinned, though, leading to a reduced scope/usefulness of the current implementation.

**Reply by SchroederSteffen** [reply]
> All others are pinned, though, leading to a reduced scope/usefulness of the current implementation.

Well, since the mentioned case is only about vulnerable direct dependencies, it's already quite limited anyways.
I was actually looking for transitive remediations, since most vulnerabilities are reported for transient dependencies.

**Reply by HonkingGoose** [reply]
> I was confused by the [documentation](https://docs.renovatebot.com/configuration-options/#transitiveremediation):

I'm improving the docs with this PR:

- #28732

**Reply by SchroederSteffen** [reply]
Many thanks for your effort @HonkingGoose, even if it was in vain. 👍

**Reply by HonkingGoose** [reply]
Thank you for the kind words! :heart: 

We're dropping the feature soon, so we don't need the docs improvement either. :upside_down_face:

**Reply by joevin-slq-docto** [reply]
Since `transitiveRemediation` has been dropped, does renovate plan to handle CVEs affecting transitive npm dependencies ?

**Reply by gunzy83** [reply]
I just ran into this today after messing with our repos trying to get a PR to trigger and then attempting on my test repos that I use to config testing. The logs show it is clearly finding the vulnerability alerts but nothing is being raised because it is a transitive dependency.

Is there any plans to support this? It would be nice to not need to have dependabot updates configured as well.

**Reply by karlhorky** [reply]
For the case Renovate doesn't support security updates for transitive deps for a while, I also opened an issue for full Dependabot support of pnpm transitive updates (this currently fails):

- https://github.com/dependabot/dependabot-core/issues/13177
- https://bsky.app/profile/did:plc:b3xwbnwkl5qzjjhkwoimvb6x/post/3lznzkrwer22a
- https://x.com/karlhorky/status/1971199721206427910
